src/main/java/edu/wpi/grip/core/Socket.java

```diff
@@ -96,19 +98,36 @@ public T getValue() {
     }
 
     /**
-     * @param step The step that this socket is part of.
+     * @param step The step that this socket is part of, if it's in a step.
      */
-    public void setStep(Optional<Step> step) {
+    protected void setStep(Optional<Step> step) {
+        step.ifPresent(s -> checkState(!this.source.isPresent(), "Socket cannot be both in a step and a source"));
         this.step = step;
     }
 
     /**
-     * @return The step that this socket is part of, or <code>null</code> if it has not been set.
+     * @return The step that this socket is part of
+     * @see #getSource()
      */
     public Optional<Step> getStep() {
         return step;
     }
 
+    /**
+     * @param source The source that this socket is part of, if it's in a source.
+     */
+    protected void setSource(Optional<Source> source) {
```

Comment:
Is there some common interface between Source and Step we could use so that we don't need to have two optional fields?
That would let socket always have an "owner"